### PR TITLE
User Story: Add Available Dates Upon Listing

### PR DESCRIPTION
**Pull request by ollienorman**
Issue #10

### CODE REVIEW

File: db/migrations/03_setting_up_available_nights_table.sql [hunk 1]
```diff
@@ -0,0 +1 @@
+CREATE TABLE available_nights (id SERIAL PRIMARY KEY, date DATE, listing_id INTEGER REFERENCES listings (id));
```
Comment (Will-Helliwell) [inline db/migrations/03_setting_up_available_nights_table.sql]: We think that the listings::available_dates relationship is a many-to-many. This table is set up as a many-to-one (which will cause date duplication in the available_dates table if, for example, you wanted to have multiple properties available on the same date). Perhaps a solution would be to use a join table and so have three in total: listings, dates/calendar, join table (listings_available-dates). We have drawn a little diagram of this in excel if it's more clear and would help!

This article goes through a similar example and at the bottom shows sql commands for accessing information using INNER JOIN command: https://learn.co/lessons/sql-join-tables-readme

File: views/listings.erb [hunk 2]
```diff
@@ -2,7 +2,7 @@
   <% @listings.each do |listing| %>
     <li>
       <%= listing.id %> <br>
```
Comment (Will-Helliwell) [inline views/listings.erb]: perhaps don't need to display the id to the customer?

File: lib/listing.rb [hunk 3]
```diff
@@ -3,25 +3,34 @@
...
+    result_available_nights = conn.exec("INSERT INTO available_nights (listing_id, date) VALUES ('#{result_listings[0]['id']}', '#{available_night}') RETURNING date;")
+    Listing.new(id: result_listings[0]['id'], title: result_listings[0]['title'], description: result_listings[0]['description'], price: result_listings[0]['price'].to_i, available_night: result_available_nights[0]['date'])
   end
 
   def self.all
     conn = connect_to_database
-    result = conn.exec("SELECT * FROM listings")
-    result.map { |listing| Listing.new(id: listing['id'], title: listing['title'], description: listing['description'], price: listing['price'].to_i) }
+    result = conn.exec("SELECT * FROM listings JOIN available_nights ON listings.id=available_nights.listing_id")
+    p result[0]['listing_id']
```
Comment (Will-Helliwell) [inline lib/listing.rb]: A few p statements that come up when you run RSPEC that could be removed

File: db/migrations/03_setting_up_available_nights_table.sql (hunk 1 above)
Comment (calavell) [reply db/migrations/03_setting_up_available_nights_table.sql]: if you could share the diagram please

File: views/listings.erb (hunk 2 above)
Comment (calavell) [reply views/listings.erb]: removed the id

File: lib/listing.rb (hunk 3 above)
Comment (calavell) [reply lib/listing.rb]: removed

File: spec/features/new_listing_spec.rb [hunk 4]
```diff
@@ -1,16 +1,27 @@
...
     expect(page).to have_content 'testtitle'
     expect(page).to have_content 'testtitle2'
     expect(page).not_to have_content 'sinatra'
   end
+
+  scenario "can add available dates" do
+    add_listing1
+    click_link('testtitle')
+    expect(page).to have_content 'testtitle'
+    expect(page).to have_content '01/01/2021'
```
Comment (Will-Helliwell) [inline spec/features/new_listing_spec.rb:29]: Nothing to change for now, but perhaps a 'nice to have' which we can incorporate into the work plan: not being able to select dates from the past

**Review by Will-Helliwell — COMMENTED** [review]
Really good guys, got it up and running in my machine and worked well. Like the drop down date selector

File: lib/listing.rb [hunk 5]
```diff
@@ -3,25 +3,34 @@
...
+    @available_night = available_night
   end
 
-  def self.create(title:, description:, price:)
+  def self.create(title:, description:, price:, available_night:)
     conn = connect_to_database
-    result = conn.exec("INSERT INTO listings (title, description, price) VALUES ('#{title}', '#{description}', '#{price}') RETURNING id, title, description, price;")
-    Listing.new(id: result[0]['id'], title: result[0]['title'], description: result[0]['description'], price: result[0]['price'].to_i)
+    result_listings = conn.exec("INSERT INTO listings (title, description, price) VALUES ('#{title}', '#{description}', '#{price}') RETURNING id, title, description, price;")
+    result_available_nights = conn.exec("INSERT INTO available_nights (listing_id, date) VALUES ('#{result_listings[0]['id']}', '#{available_night}') RETURNING date;")
```
Comment (xavierloos) [inline lib/listing.rb]: I am not sure if this is right (to have two connect in one method), perhaps you could create another method with this connection and then call it in `self.create`. Just wondering as this method is a bit long.

File: app.rb [hunk 6]
```diff
@@ -46,12 +46,8 @@ class AbodenB < Sinatra::Base
   end
 
   post "/listings/new" do
```
Comment (xavierloos) [inline app.rb:55]: As a non-user I can type the URL and it let me in, add a condition.

Comment (calavell) [reply app.rb:55]: fixed by Will & Javi

**Pull request merged**